src/lib/cache/swr.ts

```diff
@@ -0,0 +1,82 @@
+type CacheEntry<T> = {
+  data: T;
+  validUntil: number;
+};
+
+type SWROptions = {
+  ttlMs?: number; // How long until data is considered stale
+  maxEntries?: number; // Max number of entries to keep in cache
+};
+
+export class SWRCache<K, V> {
+  private cache = new Map<K, CacheEntry<V>>();
+  private inFlight = new Map<K, Promise<V>>();
+  private readonly options: Required<SWROptions>;
+
+  constructor(options: SWROptions = {}) {
+    this.options = {
+      ttlMs: options.ttlMs ?? 5 * 60 * 1000, // 5 minutes default
+      maxEntries: options.maxEntries ?? 1000,
+    };
+  }
+
+  async get(key: K, fetchFn: () => Promise<V>): Promise<V> {
+    const entry = this.cache.get(key);
+    const now = Date.now();
+
+    // If no cached data, handle fetch with deduplication
+    if (!entry) {
+      return this.dedupedFetch(key, fetchFn);
+    }
+
+    // Check if stale
+    const isStale = now > entry.validUntil;
+
+    // If stale, trigger background revalidation and return stale data
+    if (isStale) {
+      this.dedupedFetch(key, fetchFn).catch(() => {
+        // Silence background revalidation errors
+      });
```

Comment:
So this catch makes it if any error occurs while fetching, the old data continues to be used.

Let's at least log a warning here. If this errors silently we won't know if the cache is stuck in a stale state.